src/net/i2p_address.cpp

```diff
@@ -0,0 +1,200 @@
+        if (!port.empty() && !epee::string_tools::get_xtype_from_string(porti, std::string{port}))
+            return {net::error::invalid_port};
+
+        static_assert(b32_length + sizeof(tld) == sizeof(i2p_address::host_), "bad internal host size");
+        return i2p_address{host, porti};
+    }
+
+    bool i2p_address::_load(epee::serialization::portable_storage& src, epee::serialization::section* hparent)
+    {
+        i2p_serialized in{};
+        if (in._load(src, hparent) && in.host.size() < sizeof(host_) && (in.host == unknown_host || !host_check(in.host).has_error()))
+        {
+            std::memcpy(host_, in.host.data(), in.host.size());
+            std::memset(host_ + in.host.size(), 0, sizeof(host_) - in.host.size());
+            port_ = in.port;
+            return true;
+        }
+        static_assert(sizeof(unknown_host) <= sizeof(host_), "bad buffer size");
+        std::memcpy(host_, unknown_host, sizeof(unknown_host)); // include null terminator
+        port_ = 0;
+        return false;
+    }
+
+    bool i2p_address::store(epee::serialization::portable_storage& dest, epee::serialization::section* hparent) const
+    {
+        const i2p_serialized out{std::string{host_}, port_};
+        return out.store(dest, hparent);
+    }
+
+    i2p_address::i2p_address(const i2p_address& rhs) noexcept
+      : port_(rhs.port_)
+    {
+        std::memcpy(host_, rhs.host_, sizeof(host_));
+    }
+
+    i2p_address& i2p_address::operator=(const i2p_address& rhs) noexcept
+    {
+        if (this != std::addressof(rhs))
+        {
+            port_ = rhs.port_;
+            std::memcpy(host_, rhs.host_, sizeof(host_));
+        }
+        return *this;
+    }
+
+    bool i2p_address::is_unknown() const noexcept
+    {
+        static_assert(1 <= sizeof(host_), "host size too small");
+        return host_[0] == '<'; // character is not allowed otherwise
+    }
+
+    bool i2p_address::equal(const i2p_address& rhs) const noexcept
+    {
+        return port_ == rhs.port_ && is_same_host(rhs);
+    }
+
+    bool i2p_address::less(const i2p_address& rhs) const noexcept
+    {
+        return std::strcmp(host_str(), rhs.host_str()) < 0 || port() < rhs.port();
+    }
+
+    bool i2p_address::is_same_host(const i2p_address& rhs) const noexcept
+    {
+        return std::strcmp(host_str(), rhs.host_str()) == 0;
+    }
+
+    std::string i2p_address::str() const
+    {
+        const std::size_t host_length = std::strlen(host_str());
+        const std::size_t port_length =
+            port_ == 0 ? 0 : std::numeric_limits<std::uint16_t>::digits10 + 2;
+
+        std::string out{};
+        out.reserve(host_length + port_length);
+        out.assign(host_str(), host_length);
+
+        if (port_ != 0)
+        {
+            out.push_back(':');
+            namespace karma = boost::spirit::karma;
+            karma::generate(std::back_inserter(out), karma::ushort_, port());
+        }
+        return out;
+    }
+}
```

Comment:
This looks very much like the existing tor one. Can it be mostly merged ?

Reply:
I considered that but their are other i2p addressing schemes too, so keeping separate should offer flexibility in adding those later on.